metals/src/main/scala/scala/meta/internal/bsp/BspConfigGenerator.scala

```diff
@@ -0,0 +1,69 @@
+package scala.meta.internal.bsp
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.internal.bsp.BspConfigGenerationStatus._
+import scala.meta.internal.builds.BuildServerProvider
+import scala.meta.internal.builds.BuildTool
+import scala.meta.internal.builds.BuildTools
+import scala.meta.internal.builds.ShellRunner
+import scala.meta.internal.metals.Messages.BspProvider
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.MetalsLanguageClient
+import scala.meta.io.AbsolutePath
+
+import org.eclipse.lsp4j.MessageActionItem
+
+/**
+ * Runs a process to create a .bsp entry for a givev buildtool.
+ */
+final class BspConfigGenerator(
+    workspace: AbsolutePath,
+    languageClient: MetalsLanguageClient,
+    buildTools: BuildTools,
+    shellRunner: ShellRunner
+)(implicit ec: ExecutionContext) {
+  def runUnconditionally(
+      buildTool: BuildTool,
+      args: List[String]
+  ): Future[BspConfigGenerationStatus] =
+    shellRunner
+      .run(
+        s"${buildTool.executableName} bspConfig",
+        args,
+        workspace,
+        buildTool.redirectErrorOutput
+      )
+      .map(BspConfigGenerationStatus.fromExitCode)
+
+  /**
+   * Given multiple build tools that are all BuildServerProviders, allow the
+   * choose the desired build server and then connect to it.
+   */
+  def chooseAndGenerate(
+      buildTools: List[BuildTool with BuildServerProvider]
+  ): Future[(BuildTool, BspConfigGenerationStatus)] = {
+    for {
+      Some(buildTool) <- chooseBuildServerProvider(buildTools)
```

Comment:
🤔 yea true. I'll have to play with this a bit. This will only _ever_ be hit if there are two bsp server available to generate apart from bloop. At the moment, that impossible, since we don't have two BuildTools that extend `BuildServerProvider`, but still good to check.